### ISSUE TITLE
Enforcing group / key_share order in TLS1.3

### ISSUE THREAD

**Issue by sebastianas** [openssl/openssl#10832]
gnutls-cli sends by default (in the supported groups extension) _secp256r1_ first and later _x25519_. The key_share extension contains a key for both types. The server has both groups and _x25519_ comes first.
The handshake however ends up with _secp256r1_.

It would be nice to tell openssl to prefer _x25519_ over _secp256r1_.
Preferably if this could also be configured system wide via openssl.cnf in a way like:
- First supported group with a key sent by the client (what we have now).
- First supported group by the server which is also supported by the client and has a key share
- If I understand the specification correct then there could be a third case in which the client supports a group which did not contain key material and the server could select that one leading to an extended handshake.

Does this sound reasonable?

**Comment by mattcaswell** [comment]
> If I understand the specification correct then there could be a third case in which the client supports a group which did not contain key material and the server could select that one leading to an extended handshake.

Yes. We will do that at the moment if the client's key_share is for a group that is unacceptable to the server, but the client does have another supported group that would be acceptable. In that case we will issue an HRR and request the client send back a key_share for that group.

What we don't support in that scenario is where the offered key_share is for an acceptable group, but the client has a supported group which is preferable to the server. In that case we might want the server to force an HRR anyway.

So, (restating your bullets slightly differently) 3 strategies:

1) Select the first key_share sent by the client which is acceptable to the server (default)
2) Select the key_share which is the highest in the server's preference list and the client has supplied that key_share in its initial ClientHello
3) Choose the group that is highest in the server's preference order and is also supported by the client. Issue an HRR to get the key_share if necessary

**Comment by ondrejholas** [comment]
Strategy 3 (as mentioned above) can be achieved by adding a check whether existing key_share's group is the most preferred common group and if not, treat it the same way as unsupported by the server.

Attached patch adds such a check, but it is incomplete - key_share acceptance strategy should depend on SSL_OP_CIPHER_SERVER_PREFERENCE or (better) have its own tuning flag, since issuing HRR has also performance impact and unnecessary rejection of key_share may not be desirable even if server order of cipher suites is enforced.

[openssl-1.1.1k-tls13_force_server_order_of_groups_in_key_share-20210626-1356.txt](https://github.com/openssl/openssl/files/6720059/openssl-1.1.1k-tls13_force_server_order_of_groups_in_key_share-20210626-1356.txt)

**Comment by t8m** [comment]
Since the 3.5 version the new TLS group configuration allows for such enforcement.
